linkerd/proxy/spire-client/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+#![deny(rust_2018_idioms, clippy::disallowed_methods, clippy::disallowed_types)]
+#![forbid(unsafe_code)]
+
+mod api;
+
+pub use api::{Api, SvidUpdate};
+use linkerd_error::{Error, Result};
+use linkerd_identity::Credentials;
+use linkerd_identity::Id;
+use std::fmt::Display;
+use std::time::{Duration, UNIX_EPOCH};
+use tokio::sync::watch;
+use tower::Service;
+use tracing::error;
+
+pub struct Spire {
+    id: Id,
+}
+
+// === impl Spire ===
+
+impl Spire {
+    pub fn new(id: Id) -> Self {
+        Self { id }
+    }
+
+    pub async fn run<C, S>(self, credentials: C, mut client: S)
+    where
+        C: Credentials,
+        S: Service<(), Response = tonic::Response<watch::Receiver<SvidUpdate>>>,
+        S::Error: Into<Error> + Display,
+    {
+        match client.call(()).await {
+            Ok(rsp) => consume_updates(&self.id, rsp.into_inner(), credentials).await,
+            Err(error) => error!(%error, "could not establish SVID stream"),
+        }
+    }
+}
+
+async fn consume_updates<C>(
+    id: &Id,
+    mut updates: watch::Receiver<api::SvidUpdate>,
+    mut credentials: C,
+) where
+    C: Credentials,
+{
+    loop {
+        let svid_update = updates.borrow_and_update().clone();
+        if let Err(error) = process_svid(&mut credentials, svid_update, id) {
+            tracing::error!("Error processing SVID update: {}", error);
+        }
+
+        if let Err(error) = updates.changed().await {
+            tracing::debug!("SVID watch closed; terminating {}", error);
+            return;
+        }
+    }
+}
+
+fn process_svid<C>(credentials: &mut C, mut update: SvidUpdate, id: &Id) -> Result<()>
+where
+    C: Credentials,
+{
+    if let Some(svid) = update.svids.remove(id) {
+        use x509_parser::prelude::*;
+
+        let (_, parsed_cert) = X509Certificate::from_der(&svid.leaf.0)?;
+        let exp: u64 = parsed_cert.validity().not_after.timestamp().try_into()?;
+        let exp = UNIX_EPOCH + Duration::from_secs(exp);
+
+        return credentials.set_certificate(svid.leaf, svid.intermediates, svid.private_key, exp);
+    }
+
+    Err("could not find an SVID".into())
```

Comment:
We tend not to use string errors, instead using typed errors via the thiserror crate.
```rust

#[derive(Debug, thiserror::Error)]
#[error("no matching SVID found")]
pub struct NoMatchingSVIDFound(());
```

Note that the unit field ensures the error can't be built outside of the module.

```suggestion
    Err(NoMatchingSVIDFound(()).into())
```